samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -272,4 +285,44 @@ StreamProcessor createStreamProcessor(
           taskFactory.getClass().getCanonicalName()));
     }
   }
+
+  /**
+   * Generates a unique latch id that is consistent across processors within the same application lifecycle.
+   * Each {@code StreamSpec} has an id that is unique and is used for hashcode computation.
+   *
+   * @param intStreams list of {@link StreamSpec}s
+   * @return latch id
+   */
+  private String generateLatchId(List<StreamSpec> intStreams) {
+    return String.valueOf(Objects.hashCode(intStreams.stream()
+        .map(StreamSpec::getId)
+        .collect(Collectors.toList())));
+  }
+
+  /**
+   * In order to fix SAMZA-1385, we are limiting the scope of coordination util within stream creation phase and destroying
+   * the coordination util right after. By closing the zk connection, we clean up the ephemeral node used for leader election.
+   * It results in a leader election whenever a new process joins and also creates intermediate streams.
+   *    1. It is unnecessary to re-conduct leader election for stream creation in the same application lifecycle
+   *    2. Underlying systems may not support check for stream existence prior to creation which could have potential problems.
+   * As in interim solution, we reuse the same latch as a marker to determine if create streams phase is done for the
+   * application lifecycle using {@link Latch#await(long, TimeUnit)}
+   *
+   * @param streamCreationLatch latch used for stream creation
+   * @return true if processor needs to be part of election
+   *         false otherwise
+   */
+  private boolean shouldContestInElectionForStreamCreation(Latch streamCreationLatch) {
+    boolean eligibleForElection = true;
+
+    try {
+      streamCreationLatch.await(LEADER_ELECTION_WAIT_TIME_MS, TimeUnit.MILLISECONDS);
```

Comment:
The usage of the `latch` interface is super confusing because it blocks on await with 2 different timeouts - `LEADER_ELECTION_WAIT_TIME` and `LATCH_TIMEOUT_MINUTES` . Iiuc, we do this due to a lack of interface support to *test the status* of the latch. please correct me if I am wrong. 

I find this acceptable only because this is a stop-gap solution until we fix the coordinatorUtils interface.